### PR TITLE
feat: add support for grading mindmap components using submissions API

### PR DESCRIPTION
**Pull request by mariajgrimaldi**
### Description
This PR adds support for grading mind-map components based on the submissions API. We draw inspiration from the [staff graded assignments xblock](https://github.com/mitodl/edx-sga/blob/master/edx_sga/sga.py) design.

To sup up the changes:
- First we configured the component as graded, adding `has_score` and the necessary configurations for the xblock like `weight` and `points`
- After we differentiated between `saving` and `submitting`, now the former saves the mindmap in the block state and the latter creates the submission using the API
- Then, we created the necessary handlers to manage grades: `enter_grade`, `remove_grade` including their helper methods
- We connected the handlers to the UI
- Finally, we created a way for easily testing the assignment of the grade by modifying the xblock templates, see `How to test` for a detailed explanation

A bit about the design:
- Just instructors (course staff) can manage grades
- A student can submit their assignment if and only if: the due date has not passed (configured in the courses' subsection),  the student has not already submitted an assignment for that block, and the student has not already received a score. If one of these conditions is not met, then the buttons `Save assignment` or `Submit` should not appear
- A student can save their assignment any number of times

### How to test
1. Install this xblock's version in your dev environment
2. Create a new course
3. Create a new `Section > Subsection > Unit > Component: Mindmap`
4. You can use these configurations:
5. Add a simple grading configuration to your course:

Your course > Settings > Grading

Your course > Edit subsection

#### Test cases
As a student
- Save your mindmap how many times you want
- Submit your mindmap, reload the page. After reloading the buttons `Save` or `Submit` should not be available

[Screencast from 18-08-23 17:17:22.webm](https://github.com/eduNEXT/xblock-mindmap/assets/64440265/a45dc612-bbc1-4f62-b0fa-cf4fd97c632b)

Then, as an instructor:
- Go to your components, and you'll see a few buttons and inputs:

- Press `Get submissions` to get all the submissions for the current block, then check the dev console:
image

- Enter a grade for the student that submitted the mindmap, by pressing `Submit` with the submission ID and the grade into the inputs:
image

As a student:
- Reload your component, you should see the grade the instructor gave you
image

Again, as an instructor:
- Reset the score, by pressing `Remove grade` after filling the student ID you copied from the dev console:
image

As a student:
- You shouldn't see the score after reloading the page.

Now, from the components' staff debug, delete the student state:
image

Then, as a student, reload the component again (you have no data saved since we deleted it in our previous step).

Now, configure a due date way back in the past for the subsection:
image

As a student, enter the component:
image

You'll see that you can't submit the mindmap because the due date passed. Now, as an instructor, add an individual date extension:
Instructor dashboard > Extensions
image

Then, as a student, submit a new mindmap.

### Questions not resolved yet
- Can a student re-submit their assignment after an instructor resets their score?

### CODE REVIEW

File: mindmap/static/html/mindmap.html [hunk 1]
```diff
@@ -9,7 +9,35 @@
...
+        <input id="submission-id-input_{{xblock_id}}" name="submission_id"/>
+        <label for="submission-id-input_{{xblock_id}}"> Submission ID </label>
+        <input id="grade-input_{{xblock_id}}" name="grade"/>
+        <label for="grade-input_{{xblock_id}}"> Grade </label>
+        <input id="student-id-input_{{xblock_id}}" name="student_id"/>
+        <label for="student-id-input_{{xblock_id}}"> Student ID </label>
+        <button id="enter-grade-button_{{xblock_id}}" type="submit">{% trans "Submit" %}</button>
+        <button id="remove-grade-button_{{xblock_id}}" type="button"> {% trans "Remove grade" %} </button>
+    </div>
+</div>
```
Comment (mariajgrimaldi) [inline mindmap/static/html/mindmap.html:41]: These changes are only for easy testing! Don't review

**Comment by mariajgrimaldi** [comment]
Separate PR fixing tests: https://github.com/eduNEXT/xblock-mindmap/pull/14

**Comment by mariajgrimaldi** [comment]
These changes were already deployed to our staging environment for mindmaps! Could you help me with testing, please? @eduNEXT/athena-team

File: mindmap/mindmap.py [hunk 2]
```diff
@@ -2,27 +2,41 @@
 
 from __future__ import annotations
```
Comment (mariajgrimaldi) [inline mindmap/mindmap.py:4]: Here are the substantial changes! If you get lost in all the changes, most are miscellaneous.

File: mindmap/static/html/mindmap.html [hunk 3]
```diff
@@ -9,7 +9,35 @@
...
+        <button id="save_button_{{xblock_id}}">{% trans "Save assignment" %}</button>
+        <button id="submit_button_{{xblock_id}}">{% trans "Submit" %}</button>
     {% endif %}
+    {% if score  %}
+        <p>{% trans "Your score is:" %} {{ score }}.</p>
+    {% endif %}
+{% endif %}
+
+
+{% if is_instructor %}
```
Comment (mariajgrimaldi) [inline mindmap/static/html/mindmap.html:22]: 📣 These changes are only for easy testing! Don't review from line 22-41

File: mindmap/static/js/src/mindmap.js [hunk 4]
```diff
@@ -22,23 +28,68 @@ function MindMapXBlock(runtime, element, context) {
...
         $(element).find(`#save_button_${context.xblock_id}`).click(function () {
-            saveMindMap(runtime, element, currentMindMap, "student");
+            handleMindMap(runtime, element, currentMindMap, saveMindMapURL);
+        });
+
+        $(element).find(`#submit_button_${context.xblock_id}`).click(function () {
+            handleMindMap(runtime, element, currentMindMap, submitMindMapURL);
+        });
+
+        $(element).find(`#get_submissions_button_${context.xblock_id}`).click(function () {
```
Comment (mariajgrimaldi) [inline mindmap/static/js/src/mindmap.js:38]: 📣 These changes are only for easy testing! Don't review from line 38-77

**Review by Ian2012 — APPROVED** [review]

**Comment by BryanttV** [comment]
Doing some tests I only have a few comments:
1. When the student submits the mind map, or the submission date has already expired, you can still edit the mind map. I think we should block editing.
2. The instructor can currently assign a grade higher than the one defined in the `points` field. Shouldn't this be checked?

Otherwise, it looks good to me!

**Comment by mariajgrimaldi** [comment]
@BryanttV: thanks for pointing that out. Can you test again? Thanks!

File: mindmap/mindmap.py [hunk 5]
```diff
@@ -230,26 +303,314 @@ def get_current_mind_map(self) -> dict:
...
+        self.points = points
+
+        weight = data.get("weight", self.weight)
+        if weight:
+            try:
+                weight = float(weight)
+            except ValueError as exc:
+                raise JsonHandlerError(400, "Weight must be a decimal number") from exc
+            if weight < 0:
+                raise JsonHandlerError(400, "Weight must be a positive decimal number")
```
[the remark this answers is not in the document]
Comment (mariajgrimaldi) [reply mindmap/mindmap.py]: yes! I did so. Thanks!

**Comment by BryanttV** [comment]
I have already tested it again. The grade is now being checked correctly!

It is still allowing the student to edit the map, as I pointed out in my first comment. Let's leave it like this?

**Comment by mariajgrimaldi** [comment]
@BryanttV: you're right, the student can still edit the mindmap, but they can't save it. Can they?

**Comment by BryanttV** [comment]
@mariajgrimaldi Yes, it is true, it cannot be saved. I think it's just a usability detail, but it doesn't affect the functionality.

**Comment by mariajgrimaldi** [comment]
@BryanttV: I'll do it in a different PR.

**Pull request merged**